### PR TITLE
enhancement/enhance random stream

### PR DESCRIPTION
**Pull request by lizhou1111**
this enhancement is about #103 .
1. Let events be evenly distributed over one second.You can specify the time interval.For example CREATE RANDOM STREAM rand_stream(i int default rand()%5) SETTINGS eps=1000, interval_time=100; means generating equal amount of data for each interval(eps / (1000 / interval_time)).
2. fixed the bug when creating random stream with unknown settings the proton-server will core-dump
3. add random stream smoke test

### CODE REVIEW

File: src/Storages/Streaming/StorageRandom.cpp [hunk 1]
```diff
@@ -556,9 +576,16 @@ void registerStorageRandom(StorageFactory & factory)
         {
             storage_random_settings->loadFromQuery(*args.storage_def);
         }
-
+        if (storage_random_settings->eps.value == 0 || storage_random_settings->interval_time.value == 0
```
Comment (yl-lisen) [inline src/Storages/Streaming/StorageRandom.cpp]: Add a empty line

Comment (yl-lisen) [inline src/Storages/Streaming/StorageRandom.cpp]: `eps == 0`, means no limit

Comment (yl-lisen) [reply src/Storages/Streaming/StorageRandom.cpp]: we can move this settings validate logic into the ctor of StorageRandom

File: tests/stream/test_stream_smoke/0021_random_stream.json [hunk 2]
```diff
@@ -254,6 +254,46 @@
...
+            "id": 12,
+            "tags": ["create random"],
+            "name": "test random stream generating function",
+            "description": "create random stream with default generating function, limit the amount of data produced per window, set interval_time, cannot be diveded by 1000ms, check the the number of data produced per second is correct;",
+            "steps":[
+              {
+                "statements": [
+                  {"client":"python", "query_type": "table", "wait":2, "query": "drop stream if exists test22_create_random"},
+                  {"client":"python", "query_type": "table", "wait":2, "query": "create random stream test22_create_random(id int default rand()%4) engine Random() settings eps=1000, interval_time=80"},
+                  {"client":"python", "query_type": "table", "query_id":"2220", "wait":2, "query":"select cnt - lag(cnt, 1) as diff from (select count(1) as cnt from test22_create_random limit 2 emit periodic 1s) offset 1 settings max_threads=10"}
```
Comment (yl-lisen) [inline tests/stream/test_stream_smoke/0021_random_stream.json:288]: Add a case for testing the count per 80ms (i.e. interval_time)

Comment (yl-lisen) [reply tests/stream/test_stream_smoke/0021_random_stream.json:288]: Add exception cases: interval time  `==0` or `> 1000`

File: src/Storages/Streaming/StorageRandom.cpp [hunk 3]
```diff
@@ -393,21 +403,25 @@ class GenerateRandomSource final : public ISource
 
 protected:
     Chunk generate() override
-    {
+    {
```
Comment (chenziliang) [inline src/Storages/Streaming/StorageRandom.cpp]: Extra spaces

File: src/Storages/Streaming/StorageRandom.cpp (hunk 1 above)
Comment (chenziliang) [inline src/Storages/Streaming/StorageRandom.cpp]: When user don't set an `eps` or the `eps = 0`, it shall mean no limit (generate as fast as possible) ? This shall be a valid configuration ?

File: src/Storages/Streaming/StorageRandom.cpp [hunk 4]
```diff
@@ -610,12 +637,14 @@ Pipe StorageRandom::read(
     /// Will create more seed values for each source from initial seed.
     pcg64 generate(random_seed);
     
+
```
Comment (chenziliang) [inline src/Storages/Streaming/StorageRandom.cpp:651]: Extra empty line

File: src/Storages/Streaming/StorageRandom.h [hunk 5]
```diff
@@ -5,6 +5,7 @@
 #include <Core/Settings.h>
 #include <Storages/IStorage.h>
 #include <base/shared_ptr_helper.h>
+#include "base/types.h"
```
Comment (chenziliang) [inline src/Storages/Streaming/StorageRandom.h]: `<base/types.h`>

File: src/Storages/Streaming/StorageRandom.h [hunk 6]
```diff
@@ -13,7 +14,8 @@ namespace DB
 class ASTStorage;
 
 #define STORAGE_RANDOM_RELATED_SETTINGS(M) \
-    M(UInt64, eps, 1000, "Limits how many rows of data can be generated per window by random storages total streams", 0)
+    M(UInt64, eps, 1000, "Limits how many rows of data can be generated per second by random storages per stream (each query will generate $num_streams source by default). 0 means no maximum value", 0) \
```
Comment (chenziliang) [inline src/Storages/Streaming/StorageRandom.h]: `Limit how many rows to be generated per second for each thread. Used by RANDOM STREAM. 0 means no limit.`

File: src/Storages/Streaming/StorageRandom.h [hunk 7]
```diff
@@ -13,7 +14,8 @@ namespace DB
...
+    M(UInt64, interval_time, 100, "the data generating interval, unit ms", 0)
```
Comment (chenziliang) [inline src/Storages/Streaming/StorageRandom.h:17]: When you have `eps`, why you need `interval` ?

Comment (lizhou1111) [reply src/Storages/Streaming/StorageRandom.h:17]: when you set EPS = 2000, it means generating 2000 pieces of data per second. when you set interval_time = 200ms, it means dividing one second into equal parts of 200ms and every interval will generate 400 pieces of data.The reason for this setting is to generate data evenly within one second.

File: src/Storages/Streaming/StorageRandom.cpp (hunk 1 above)
Comment (lizhou1111) [reply src/Storages/Streaming/StorageRandom.cpp]: In issue #72, Jove suggested that I set the default value to 1000.If the user explicitly sets eps = 0, this means there is no limit

File: src/Storages/Streaming/StorageRandom.cpp [hunk 8]
```diff
@@ -370,13 +372,21 @@ class GenerateRandomSource final : public ISource
...
         partial_size = events_per_second_ % block_size;
+        /// calculate the number of interval during 1 sec
+        interval_count = 1000 / generate_interval;
+        /// special interval
+        last_interval_time = generate_interval + 1000 % generate_interval;
+        boundary_time = MonotonicMilliseconds::now() + last_interval_time;
+        /// calculate the number of data generated per interval
+        normal_interval = events_per_second / interval_count;
+        /// calculate the number of data generated in the special interval
+        last_interval_count = normal_interval + events_per_second % interval_count;
```
Comment (yl-lisen) [inline src/Storages/Streaming/StorageRandom.cpp]: It's complex code here, we should have some clearer comments, and add a example to explain this logic.

File: src/Storages/Streaming/StorageRandom.cpp [hunk 9]
```diff
@@ -503,7 +522,7 @@ void StorageRandomSettings::loadFromQuery(ASTStorage & storage_def)
         catch (Exception & e)
         {
             if (e.code() == ErrorCodes::UNKNOWN_SETTING)
-                e.addMessage("for storage " + storage_def.engine->name);
+                e.addMessage(e.what());
```
Comment (yl-lisen) [inline src/Storages/Streaming/StorageRandom.cpp]: addMessage("for random stream")

**Review by yl-lisen — APPROVED** [review]

**Pull request merged**